src/rebar_dir.erl

```diff
@@ -39,18 +40,31 @@ base_dir(State) ->
 
 %% @doc returns the directory root for build artifacts for a given set
 %% of profiles.
--spec profile_dir(rebar_dict(), [atom()]) -> file:filename_all().
+-spec profile_dir(rebar_dict(), [atom(), ...]) -> file:filename_all().
 profile_dir(Opts, Profiles) ->
-    {BaseDir, ProfilesStrings} = case [rebar_utils:to_list(P) || P <- Profiles] of
-        ["global" | _] -> {?MODULE:global_cache_dir(Opts), [""]};
-        ["bootstrap", "default"] -> {rebar_opts:get(Opts, base_dir, ?DEFAULT_BASE_DIR), ["default"]};
-        ["default"] -> {rebar_opts:get(Opts, base_dir, ?DEFAULT_BASE_DIR), ["default"]};
-        %% drop `default` from the profile dir if it's implicit and reverse order
+    BasePath =
+        case Profiles of
+            [global | _] -> ?MODULE:global_cache_dir(Opts);
+            [_|_] -> rebar_opts:get(Opts, base_dir, ?DEFAULT_BASE_DIR)
+        end,
+    DirName = profile_dir_name(Profiles),
+    filename:join(BasePath, DirName).
+
+%% @doc returns the directory name for build artifacts for a given set
+%% of profiles.
+-spec profile_dir_name([atom(), ...] | rebar_state:t()) -> file:filename_all().
+profile_dir_name(Profiles)
+  when is_list(Profiles) ->
+    case [rebar_utils:to_list(P) || P <- Profiles] of
+        ["global" | _] -> "";
+        ["bootstrap", "default"] -> "default";
+        ["default"] -> "default";
+        %% drop `default' from the profile dir if it's implicit and reverse order
         %%  of profiles to match order passed to `as`
-        ["default"|Rest] -> {rebar_opts:get(Opts, base_dir, ?DEFAULT_BASE_DIR), Rest}
-    end,
-    ProfilesDir = rebar_string:join(ProfilesStrings, "+"),
-    filename:join(BaseDir, ProfilesDir).
+        ["default"|NonDefaultNames] -> rebar_string:join(NonDefaultNames, "+")
+    end;
+profile_dir_name(State) ->
+    profile_dir_name(rebar_state:current_profiles(State)).
```

Comment:
Not liking this last clause. I'd rather we have `profile_dir_name(State)` and maybe something like `profile_dir_name_(List)` that is private. That way we only expose/support one public format (passing in the state) for external caller, and the rest of the formatting is private and called by the other functions. An alternative is to always ask for the explicit list of profiles to be provided (as atoms) by all callers and leave the function public.

I'd generally consider it to be a kind of anti-pattern to have dual shared signatures like that intended for totally different callers in different contexts. It smells more of a shortcut than thought-out design.

Reply:
Fixed